### PR TITLE
Added new parameter to transform_bid_ask_long_data_to_wide and added unit tes

### PR DESCRIPTION
**Pull request by Vedanshu7**
Task #743 
- Added New parameter to function transform_bid_ask_long_data_to_wide 
- Created test case for new parameter and symbols

### CODE REVIEW

File: core/finance/bid_ask.py [hunk 1]
```diff
@@ -158,6 +158,7 @@ def transform_bid_ask_long_data_to_wide(
     bid_prefix: str = "bid_",
     ask_prefix: str = "ask_",
     final_col_format: str = "{0[0]}_l{0[1]}",
+    value_col_prefix: List[str] = ["log", "half"],
```
Comment (gpsaggese) [inline core/finance/bid_ask.py:161]: You need to pass `None` and then expand it in the code

This is known to be dangerous
https://docs.quantifiedcode.com/python-anti-patterns/correctness/mutable_default_value_as_argument.html

File: core/finance/bid_ask.py [hunk 2]
```diff
@@ -184,6 +185,7 @@ def transform_bid_ask_long_data_to_wide(
     :param final_col_format: format for the final column names, pandas' df.pivot by
         default uses column naming matrix like so: columns x values.
         E.g. {bid_size, bid_price ...} x {1, 2, 3 ...} = {bid_size1, bid_size2 ...}).
+    :param value_col_prefix: prefixes for value-related columns; default is ["log", "half"].
```
Comment (gpsaggese) [inline core/finance/bid_ask.py:188]: no need to say what's the default. That belongs to the interface

File: core/finance/test/test_bid_ask.py [hunk 3]
```diff
@@ -270,19 +270,26 @@ def get_df_with_long_levels(self) -> pd.DataFrame:
...
+            "log_size": pd.Series([0.01, 0.03, 0.01]),
         }
         df = pd.DataFrame(data=values)
         df = df.set_index("timestamp")
         return df
 
     def test1(self) -> None:
+        """
+        Test the transformation of long bid-ask data to wide format with
+        deafult agrmunet value of `value_col_prefix`.
```
Comment (gpsaggese) [inline core/finance/test/test_bid_ask.py:291]: default argument

After sending a PR, pls review it yourself top-to-bottom for imperfections

File: core/finance/bid_ask.py [hunk 4]
```diff
@@ -194,7 +196,7 @@ def transform_bid_ask_long_data_to_wide(
         col
         for col in df.columns
         # TODO(Juraj): ad-hoc hack, address properly in #CmTask7387.
-        if col.startswith((bid_prefix, ask_prefix, "log", "half"))
+        if col.startswith((bid_prefix, ask_prefix) + tuple(value_col_prefix))
```
Comment (gpsaggese) [inline core/finance/bid_ask.py:199]: This is valid but difficult to read

Let's use something more readable like
```
prefixes = [bid_prefix, ask_prefix] + list(value_col_prefix)
...
if any(col.startswith(prefix) for prefix in prefixes)

```

File: core/finance/bid_ask.py (hunk 1 above)
Comment (gpsaggese) [inline core/finance/bid_ask.py:161]: should be plural `...prefixes`.

File: core/finance/bid_ask.py (hunk 2 above)
Comment (Vedanshu7) [reply core/finance/bid_ask.py]: Resolved.

File: core/finance/bid_ask.py (hunk 1 above)
Comment (Vedanshu7) [reply core/finance/bid_ask.py]: Resolved.

File: core/finance/test/test_bid_ask.py (hunk 3 above)
Comment (Vedanshu7) [reply core/finance/test/test_bid_ask.py]: Resolved.

File: core/finance/bid_ask.py (hunk 4 above)
Comment (Vedanshu7) [reply core/finance/bid_ask.py]: Resolved.

File: core/finance/bid_ask.py (hunk 1 above)
Comment (Vedanshu7) [reply core/finance/bid_ask.py]: Resolved.

File: core/finance/bid_ask.py [hunk 5]
```diff
@@ -184,17 +185,21 @@ def transform_bid_ask_long_data_to_wide(
...
+    if value_col_prefixes is None:
+        value_col_prefixes = ["log","half"]
     hdbg.dassert_in(timestamp_col, df.reset_index().columns)
     df = df.reset_index()
+    prefixes = [bid_prefix, ask_prefix] + list(value_col_prefixes)
     # Specify bid-ask and non-bid-ask columns.
     bid_ask_cols = [
         col
         for col in df.columns
         # TODO(Juraj): ad-hoc hack, address properly in #CmTask7387.
```
Comment (jsmerix) [inline core/finance/bid_ask.py:201]: We can remove this comment now - this is the issue in which we are solving the problem

File: core/finance/test/test_bid_ask.py [hunk 6]
```diff
@@ -270,19 +270,26 @@ def get_df_with_long_levels(self) -> pd.DataFrame:
             pd.Timestamp("2022-09-08 21:01:15+00:00"),
         ]
         values = {
-            "level": [1, 2, 3],
+            "level": [1, 1, 1],
```
Comment (jsmerix) [inline core/finance/test/test_bid_ask.py:273]: Why did you change this?

**Review by jsmerix — CHANGES_REQUESTED** [review]

File: core/finance/bid_ask.py (hunk 5 above)
Comment (Vedanshu7) [reply core/finance/bid_ask.py]: Resolved.

File: core/finance/test/test_bid_ask.py (hunk 6 above)
Comment (Vedanshu7) [reply core/finance/test/test_bid_ask.py:273]: while running the one-symbol, one-level test case. The transform_bid_ask_long_data_to_wide function is throwing an error for duplicate entries, and it couldn't be to reshape; the pivot function is failing.

This is my test data:

```
                                               level currency_pair  bid_price  bid_size  \
timestamp                                                             
2022-09-08 21:01:00+00:00      1      XRP_USDT       2.31       1.1   
2022-09-08 21:01:00+00:00      1      XRP_USDT       3.22       2.2   
2022-09-08 21:01:00+00:00      1      XRP_USDT       2.33       3.3   

                                                 ask_price  ask_size       knowledge_timestamp  \
timestamp                                                                  
2022-09-08 21:01:00+00:00       2.34       5.4 2022-09-08 21:01:15+00:00   
2022-09-08 21:01:00+00:00       3.24       5.5 2022-09-08 21:01:15+00:00   
2022-09-08 21:01:00+00:00       2.35       6.6 2022-09-08 21:01:15+00:00   

                                              half_spread  log_size  
timestamp                                         
2022-09-08 21:01:00+00:00        0.005      0.01  
2022-09-08 21:01:00+00:00        0.020      0.03  
2022-09-08 21:01:00+00:00        0.020      0.01  
```

this is the error trace:

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[4], line 7
      5        #
      6 timestamp_col = "timestamp"
----> 7 wide_levels_df = cfibiask.transform_bid_ask_long_data_to_wide(
      8            long_levels_df, timestamp_col
      9 )
     11 print(wide_levels_df)

File /app/core/finance/bid_ask.py:210, in transform_bid_ask_long_data_to_wide(df, timestamp_col, bid_prefix, ask_prefix, final_col_format, value_col_prefixes)
    204 non_bid_ask_cols = [
    205     col for col in df.columns if col not in bid_ask_cols + ["level", "id"]
    206 ]
    207 # TODO(Max): Create an assertion that all values for levels are identical,
    208 # so we are merging the rows without duplicates (i.e., "knowledge_timestamp" and "end_download_timestamp").
    209 # Merge `level` into bid-ask values (e.g., bid_price_1, bid_price_2, etc.).
--> 210 pivoted_data = df.pivot(
    211     index=non_bid_ask_cols,
    212     columns=["level"],
    213     values=bid_ask_cols,
    214 )
    215 # Rename the columns to a desired {value}_{level} format.
    216 pivoted_data.columns = pivoted_data.columns.map(final_col_format.format)

File /venv/lib/python3.9/site-packages/pandas/core/frame.py:9028, in DataFrame.pivot(self, columns, index, values)
   9021 @Substitution("")
   9022 @Appender(_shared_docs["pivot"])
   9023 def pivot(
   9024     self, *, columns, index=lib.no_default, values=lib.no_default
   9025 ) -> DataFrame:
   9026     from pandas.core.reshape.pivot import pivot
-> 9028     return pivot(self, index=index, columns=columns, values=values)

File /venv/lib/python3.9/site-packages/pandas/core/reshape/pivot.py:553, in pivot(data, columns, index, values)
    549         indexed = data._constructor_sliced(data[values]._values, index=multiindex)
    550 # error: Argument 1 to "unstack" of "DataFrame" has incompatible type "Union
    551 # [List[Any], ExtensionArray, ndarray[Any, Any], Index, Series]"; expected
    552 # "Hashable"
--> 553 result = indexed.unstack(columns_listlike)  # type: ignore[arg-type]
    554 result.index.names = [
    555     name if name is not lib.no_default else None for name in result.index.names
    556 ]
    558 return result

File /venv/lib/python3.9/site-packages/pandas/core/frame.py:9625, in DataFrame.unstack(self, level, fill_value, sort)
   9561 """
   9562 Pivot a level of the (necessarily hierarchical) index labels.
   9563 
   (...)
   9621 dtype: float64
   9622 """
   9623 from pandas.core.reshape.reshape import unstack
-> 9625 result = unstack(self, level, fill_value, sort)
   9627 return result.__finalize__(self, method="unstack")

File /venv/lib/python3.9/site-packages/pandas/core/reshape/reshape.py:504, in unstack(obj, level, fill_value, sort)
    502 if isinstance(obj, DataFrame):
    503     if isinstance(obj.index, MultiIndex):
--> 504         return _unstack_frame(obj, level, fill_value=fill_value, sort=sort)
    505     else:
    506         return obj.T.stack(future_stack=True)

File /venv/lib/python3.9/site-packages/pandas/core/reshape/reshape.py:529, in _unstack_frame(obj, level, fill_value, sort)
    525 def _unstack_frame(
    526     obj: DataFrame, level, fill_value=None, sort: bool = True
    527 ) -> DataFrame:
    528     assert isinstance(obj.index, MultiIndex)  # checked by caller
--> 529     unstacker = _Unstacker(
    530         obj.index, level=level, constructor=obj._constructor, sort=sort
    531     )
    533     if not obj._can_fast_transpose:
    534         mgr = obj._mgr.unstack(unstacker, fill_value=fill_value)

File /venv/lib/python3.9/site-packages/pandas/core/reshape/reshape.py:154, in _Unstacker.__init__(self, index, level, constructor, sort)
    146 if num_cells > np.iinfo(np.int32).max:
    147     warnings.warn(
    148         f"The following operation may generate {num_cells} cells "
    149         f"in the resulting pandas object.",
    150         PerformanceWarning,
    151         stacklevel=find_stack_level(),
    152     )
--> 154 self._make_selectors()

File /venv/lib/python3.9/site-packages/pandas/core/reshape/reshape.py:210, in _Unstacker._make_selectors(self)
    207 mask.put(selector, True)
    209 if mask.sum() < len(self.index):
--> 210     raise ValueError("Index contains duplicate entries, cannot reshape")
    212 self.group_index = comp_index
    213 self.mask = mask

ValueError: Index contains duplicate entries, cannot reshape
```

Comment (Vedanshu7) [reply core/finance/test/test_bid_ask.py:273]: @jsmerix When encountering duplicate rows in the non-bid-ask columns during pivoting, the pivot function is throwing an error. Options for solutions include taking the mean, summing the data, using a specific custom aggregation function, or dropping duplicates. Can you please guide me on which option I should choose to move forward?

Comment (Vedanshu7) [reply core/finance/test/test_bid_ask.py:273]: Or if all the columns in non-bid-ask have duplicate values, then it should bypass the pivoting?

Comment (jsmerix) [reply core/finance/test/test_bid_ask.py:273]: No need to update the function in any way, the input data for the test should not include duplicate timestamp for a given symbol and level

Comment (Vedanshu7) [reply core/finance/test/test_bid_ask.py]: Resolved.

**Review by jsmerix — CHANGES_REQUESTED** [review]
All of the test methods have the same pattern:

- get data
- call function
- compare with expected

Could we try to factor out common behavior to a `_test` method and simply pass the variables?

**Comment by Vedanshu7** [comment]
> All of the test methods have the same pattern:
> 
> * get data
> * call function
> * compare with expected
> 
> Could we try to factor out common behavior to a `_test` method and simply pass the variables?

Sure

**Review by jsmerix — APPROVED** [review]

**Comment by Vedanshu7** [comment]
@gpsaggese Can you please review this pull request at your earliest convenience?

**Review by gpsaggese — COMMENTED** [review]
LGTM
